debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java

```diff
@@ -0,0 +1,166 @@
+
+/**
+ * Debezium Outbox Transform Event Router
+ *
+ * @author Renato mefi (gh@mefi.in)
+ */
+public class EventRouter<R extends ConnectRecord<R>> implements Transformation<R> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EventRouter.class);
+
+    private final ExtractField<R> afterExtractor = new ExtractField.Value<>();
+    private final RegexRouter<R> regexRouter = new RegexRouter<>();
+    private EventRouterConfigDefinition.InvalidOperationBehavior invalidOperationBehavior;
+
+    private String fieldEventId;
+    private String fieldEventKey;
+    private String fieldEventType;
+    private String fieldPayload;
+    private String fieldPayloadId;
+    private String fieldPayloadType;
+
+    private String routeByField;
+
+    @Override
+    public R apply(R r) {
+        // Ignoring tombstones
+        if (r.value() == null) {
+            LOGGER.info("Tombstone message {} ignored", r.key());
+            return null;
+        }
+
+        Struct debeziumEventValue = requireStruct(r.value(), "Detect Debezium Operation");
+        String op = debeziumEventValue.getString(Envelope.FieldName.OPERATION);
+
+        // Skipping deletes
+        if (op.equals(Envelope.Operation.DELETE.code())) {
+            LOGGER.info("Delete message {} ignored", r.key());
+            return null;
+        }
+
+        // Dealing with unexpected update operations
+        if (op.equals(Envelope.Operation.UPDATE.code())) {
+            handleUnexpectedOperation(r);
+            return null;
+        }
+
+        final R afterRecord = afterExtractor.apply(r);
+        Struct eventStruct = requireStruct(afterRecord.value(), "Read Outbox Event");
+
+        Long timestamp = debeziumEventValue.getInt64("ts_ms");
+
+        String eventId = eventStruct.getString(fieldEventId);
+        String eventType = eventStruct.getString(fieldEventType);
+        String payload = eventStruct.getString(fieldPayload);
+        String payloadId = eventStruct.getString(fieldPayloadId);
+        String payloadType = eventStruct.getString(fieldPayloadType);
+
+        Schema valueSchema = SchemaBuilder.struct()
+                .field("eventType", eventStruct.schema().field(fieldEventType).schema())
+                .field("ts_ms", debeziumEventValue.schema().field("ts_ms").schema())
+                .field("payload", eventStruct.schema().field(fieldPayload).schema())
+                .build();
+
+        Struct value = new Struct(valueSchema)
+                .put("eventType", eventType)
+                .put("ts_ms", timestamp)
+                .put("payload", payload);
+
+        R newRecord = r.newRecord(
+                eventStruct.getString(routeByField).toLowerCase(),
+                null,
+                Schema.STRING_SCHEMA,
+                defineRecordKey(eventStruct, payloadId),
+                valueSchema,
+                value,
+                r.timestamp(),
+                null
```

Comment:
The SMT from the blog post propagated the event id (_not_ aggregate id) as a message header property, allowing consumers to detect duplicates efficiently, without having to parse the payload. Do you plan to add this here, too?

Reply:
Yes, it's on my to do list, I thought about treating this one as "extra fields", default going to header but with the option to be part of the event envelope. Wdyt?

Reply:
I think just header is fine. We can always evolve it later on.

Reply:
It's fixed, but I really need your validation on the approach of "extra fields"